Support for XNNPACK max pooling operator.

AshkanAliabadi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35896 Add XNNPACK Clamp.
* #35791 Enable stateless XNNPACK linear.
* #35790 Enable stateless XNNPACK convolutions.
* **#35354 Support for XNNPACK max pooling operator.**
* #35351 Propagate input tensor names in XNNPACK backend.

Differential Revision: [D20821862](https://our.internmc.facebook.com/intern/diff/D20821862)

Commit:
Support for XNNPACK max pooling operator.

[ghstack-poisoned]

Commit:
Update on "Support for XNNPACK max pooling operator."




[ghstack-poisoned]

Commit:
Update on "Support for XNNPACK max pooling operator."




[ghstack-poisoned]

Commit:
Update on "Support for XNNPACK max pooling operator."




[ghstack-poisoned]

Commit:
Update on "Support for XNNPACK max pooling operator."




[ghstack-poisoned]

Commit:
Update on "Support for XNNPACK max pooling operator."




[ghstack-poisoned]

Commit:
Update on "Support for XNNPACK max pooling operator."




[ghstack-poisoned]

ailzhang commented on aten/src/ATen/native/xnnpack/MaxPooling.cpp:
```diff
@@ -0,0 +1,198 @@
+    dilation_,
+  };
+
+  return xnnpack::internal::available() &&
```
This condition is pretty hard to read, it'll be great if you can add a verbose comment to describe when xnnpack is used.

AshkanAliabadi replied:
Done, thanks.

Commit:
Update on "Support for XNNPACK max pooling operator."




[ghstack-poisoned]

kimishpatel commented on aten/src/ATen/native/xnnpack/MaxPooling.cpp:
```diff
@@ -0,0 +1,212 @@
+    return false;
+  }
+
+  // Stride can be legitimately empty, in which case it is to be defaulted to kernel size.
```
Is this right? Default to kernel size or to 1? I dont know the convention but just making sure.

dreiss replied:
Pytorch defaults to 1 for convolution, but kernel_size for pooling.  This matches how conv and pool are "normally" used.

kimishpatel commented on aten/src/ATen/native/xnnpack/MaxPooling.cpp:
```diff
@@ -0,0 +1,212 @@
+      (2 == parameters.kernel.size()) &&
+      (parameters.kernel[Layout::Parameter::height] > 0) &&
+      (parameters.kernel[Layout::Parameter::width] > 0) &&
+      ((parameters.kernel[Layout::Parameter::height] *
+        parameters.kernel[Layout::Parameter::width]) > 1) &&
```
Why this condition? If kernel high and width both must be integers and > 0, then this condition is implied, no?

AshkanAliabadi replied:
1x1 max pooling is not meaningful and XNNPACK prohibits that use case.  As for the check, let's read > 0 as >= 1 to make it more intuitive.  With that in mind, for any integer M and N greater than or equal to 1, MxN, 1xN, and Mx1 are all valid sizes except for 1x1 which is what this check is guarding against.

kimishpatel commented on aten/src/ATen/native/xnnpack/MaxPooling.cpp:
```diff
@@ -0,0 +1,212 @@
+      // Output Min / Max
+      (output_max > output_min) &&
+      // Output
+      (pooling_output_shape(
```
Does this function assume parameters such as kernel, padding, stride both must be symmetrical, meaning both height and width of these parameters be the same? Because we are not enforcing that anywhere else.

kimishpatel commented on aten/src/ATen/native/xnnpack/MaxPooling.cpp:
```diff
@@ -0,0 +1,212 @@
+    const float output_max) {
+  using namespace internal;
+
+  // A call to max_pool2d must have been gated by a call to use_maxpool2d, so
```
Can this be enforced somehow? That is, if you come here without going through use_maxpool2d, then it is caught? And if it is not possible then it would be good to couple these two together such that this is not possible. I guess the trouble would be the API of max_pool2d. Maybe it can return optional Tensor?

AshkanAliabadi replied:
I cannot think of a way since both these functions are pure and stateless.

dreiss replied:
I think the idea is that you would remove `use_max_pool2d` as a separate function and bring all that logic into `max_pool2d`.  That function would first perform the checks.  If they fail, it would return `nullopt`, and if they succeed, it would perform the max pool and return the result tensor.  Then the calling code would change from

```
if (can_use_xnnpack()) {
  return use_xnnpack();
}
```
to
```
optional<Tensor> ret = use_xnnpack();
if (ret) {
  return *ret;
}
```

It's not clear to me whether this would be better or worse than the current implementation.

dreiss commented on aten/src/ATen/native/xnnpack/RegisterOpContextClass.cpp:
```diff
@@ -15,6 +15,7 @@ using internal::linear::createLinearClampPrePackOpContext;
 using internal::convolution2d::createConv2dClampPrePackOpContext;
```
Looks like no change to this file.  Want to revert it?

AshkanAliabadi replied:
Will do.

dreiss commented on aten/src/ATen/native/xnnpack/MaxPooling.cpp:
```diff
@@ -0,0 +1,212 @@
+
+  // Here are the list of conditions required for this code path to be taken:
+  // * Input must be 4D CPU float tensor with no gradients.
+  // * Kernel must be a 2D IntArrayRef containing two positive numbers.
+  //   Furthermore, 1x1 kernels are neither valid nor meaningful.
```
I see you commented below that XNNPACK prohibits this.  That's fine, but I'd update this comment to explain that that is the reason for returning false in this case.

AshkanAliabadi replied:
Right, XNNPACK errors on this input.  Will improve the expand the comment with the explanation.

Commit:
Update on "Support for XNNPACK max pooling operator."




[ghstack-poisoned]

Commit:
Update on "Support for XNNPACK max pooling operator."




[ghstack-poisoned]

Commit:
Update on "Support for XNNPACK max pooling operator."


Differential Revision: [D20821862](https://our.internmc.facebook.com/intern/diff/D20821862)

[ghstack-poisoned]

Commit:
Update on "Support for XNNPACK max pooling operator."


Differential Revision: [D20821862](https://our.internmc.facebook.com/intern/diff/D20821862)

[ghstack-poisoned]

Commit:
Update on "Support for XNNPACK max pooling operator."


Differential Revision: [D20821862](https://our.internmc.facebook.com/intern/diff/D20821862)

[ghstack-poisoned]

Commit:
Update on "Support for XNNPACK max pooling operator."


Differential Revision: [D20821862](https://our.internmc.facebook.com/intern/diff/D20821862)

[ghstack-poisoned]

Commit:
Update on "Support for XNNPACK max pooling operator."


Differential Revision: [D20821862](https://our.internmc.facebook.com/intern/diff/D20821862)

[ghstack-poisoned]
